feat!: flat config + v9 support, rule unification, add TS config

jirimoravcik opened the pull request:
This PR begins a new era of linting at Apify.
It introduces a new shared eslint config. It uses the flat eslint format and works with eslint v9.

I consolidated the TypeScript config into this one, so the maintenance and updates are easier. This means we won't need `apify-eslint-config-ts` anymore.

I have tried to consolidate most of the rules from configurations we have available across our projects. If you still feel something should be changed, feel free to comment here.

Newly added/enabled rules:
[no-use-before-define](https://eslint.org/docs/latest/rules/no-use-before-define)
[no-param-reassign](https://eslint.org/docs/latest/rules/no-param-reassign)
[consistent-return](https://eslint.org/docs/latest/rules/consistent-return)
[array-callback-return](https://eslint.org/docs/latest/rules/array-callback-return)
[no-promise-executor-return](https://eslint.org/docs/latest/rules/no-promise-executor-return)
[@typescript-eslint/array-type](https://typescript-eslint.io/rules/array-type/)
[@typescript-eslint/adjacent-overload-signatures](https://typescript-eslint.io/rules/adjacent-overload-signatures)
[@typescript-eslint/no-for-in-array](https://typescript-eslint.io/rules/no-for-in-array)
[@typescript-eslint/no-implied-eval](https://typescript-eslint.io/rules/no-implied-eval)
[@typescript-eslint/no-inferrable-types](https://typescript-eslint.io/rules/no-inferrable-types)
[@typescript-eslint/no-confusing-non-null-assertion](https://typescript-eslint.io/rules/no-confusing-non-null-assertion)
[@typescript-eslint/prefer-includes](https://typescript-eslint.io/rules/prefer-includes)
[import/no-default-export](https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/no-default-export.md)

Let's first review this PR, then I'll release a beta version that can be easily tested 😉 .
After that, we'll do final touches and merge this PR. After that, we'll have to update the config across our projects. I will oversee this and create/plan tasks in GitHub for it.

metalwarrior665 commented on index.js:
```diff
@@ -1,66 +1,190 @@
+            // Require methods and fucnctions to be marked async if they return a Promise.
+            '@typescript-eslint/promise-function-async': 'error',
+            // Force the usage of "import type" for type imports.
+            '@typescript-eslint/consistent-type-imports': 'error',
```
Seems this defaults to `fixStyle: 'separate-type-imports'` which would error on `import { type CheerioCrawlingContext }` if I'm reading it right? We use that syntax in some places but I'm ok to be overturned :)

https://typescript-eslint.io/rules/consistent-type-imports/#fixstyle

Jkuzz replied:
I hate `import { type` it feels messy and ends up having several `type`s mixed between the imports as the file grows. Also apparently this is better performance 🤔

stetizu1 replied:
I don't care, but having only one type seems good. I guess this can also be autofixed.

B4nan replied:
We use `import { type` quite extensively, because we have some other rule set up in a way that we can only have one import per file/dependency, but if the fixer works, I really don't mind. 

I recall those fixers tend to screw up with whitespace, but maybe it was a different rule than this one.

fnesveda commented on package.json:
```diff
@@ -1,24 +1,26 @@
+    "globals": "^15.11.0"
   },
   "peerDependencies": {
+    "typescript-eslint": "*",
     "eslint": "*"
```
Wouldn't it be better to specify ESLint 9+ and Typescript-eslint 8+ here? I'm not sure, just asking

jirimoravcik replied:
I'm not sure as well. I see it was added like that a long time ago. https://github.com/apify/apify-eslint-config/commit/1141f28f732b8068e05bb262a81d664ec637bcb6

@B4nan what do you think is the proper approach here? thanks

B4nan replied:
yes, given the BCs in the config format, I believe we are no longer compatible with older versions, so this should be either `^9.0.0` or `>=9.0.0` if we want to allow future major versions too right ahead (but we could leave that for later once we see what they will actually break), same for the TS plugin

fnesveda commented on index.js:
```diff
@@ -1,66 +1,190 @@
+            // Forcing "use strict" is not necessary.
+            strict: 'off',
+            // We don't want to force newlines before and after curly braces.
+            'object-curly-newline': 'off',
```
I would not completely disable this, but force consistency (either no newlines at all, or before/after both braces), what do you think?
```suggestion
            'object-curly-newline': ['error', { 'consistent': true }],
```

jirimoravcik replied:
Makes sense, thanks

fnesveda commented on index.js:
```diff
@@ -1,66 +1,190 @@
+            'import/no-default-export': 'error',
+            // It's ok to not have a default export and we force named exports anyway.
+            'import/prefer-default-export': 'off',
+            'import/no-unresolved': 'off',
```
Why disable this one?

jirimoravcik replied:
It was in the original config, see https://github.com/apify/apify-eslint-config/blob/master/index.js#L29
I'd be happy to remove it. Let's see what others think.

stetizu1 replied:
I am also for removing it

jirimoravcik replied:
Removed 😎

fnesveda commented on index.js:
```diff
@@ -1,66 +1,190 @@
+                    attributes: false,
+                },
+            }],
+            // Require methods and fucnctions to be marked async if they return a Promise.
```
```suggestion
            // Require methods and functions to be marked async if they return a Promise.
```

fnesveda reviewed (changes requested):
Very cool! Just a few small points.

stetizu1 commented on index.js:
```diff
@@ -1,66 +1,190 @@
+            'no-continue': 'off',
+            'arrow-body-style': 'off',
+            'no-restricted-syntax': ['error', 'ForInStatement', 'LabeledStatement', 'WithStatement'],
+            'no-prototype-builtins': 'off',
```
How often are we actually doing this? I don't think any of our developers actually does that in non-necessary cases, but it may also be limited... 🤔

jirimoravcik replied:
True, I'll remove it

stetizu1 commented on index.js:
```diff
@@ -1,66 +1,190 @@
+            'no-restricted-syntax': ['error', 'ForInStatement', 'LabeledStatement', 'WithStatement'],
+            'no-prototype-builtins': 'off',
+            // Set max line length to 160 characters.
+            'max-len': ['error', {
```
Maybe this should be in the future solved by prettier (or something) - so not sure if to include it here? Well, we can now, but we should investigate after. There are probably more rules like that.

Otherwise like the ignores.

jirimoravcik replied:
Yeah, I'll keep it for now. I didn't want to include prettier initiative as the scope would be too large

stetizu1 commented on index.js:
```diff
@@ -1,66 +1,190 @@
+                enforceForRenamedProperties: false,
+            }],
+            // So that the functions calling return show up in stack traces in case of an error.
+            'no-return-await': 'off',
```
👍 
In some cases it is even necessary (like return in try/catch) - not sure why is this rule so popular, I would force it otherwise 😅 (it saves a "tick of time")

stetizu1 commented on index.js:
```diff
@@ -1,66 +1,190 @@
+            // It's ok to not have a default export and we force named exports anyway.
+            'import/prefer-default-export': 'off',
+            'import/no-unresolved': 'off',
+            'import/extensions': 'off',
```
This may probably be a part of ESM conversion 😅

stetizu1 commented on index.js:
```diff
@@ -1,66 +1,190 @@
+                },
+                'newlines-between': 'always',
+            }],
+            'import/no-named-as-default': 'off',
```
turning this off since we have no default exports allowed is probably for nothing

jirimoravcik replied:
yeah, makes sense, removing.

stetizu1 commented on index.js:
```diff
@@ -1,66 +1,190 @@
+            // TypeScript doesn't support imports with `.ts` extensions, and `.js` makes the rule complain.
+            'import/extensions': 'off',
+            // Allow semicolons at the end of the code block.
+            '@typescript-eslint/no-extra-semi': 'off',
```
I would turn this on 🙊 Autofix would probably fix all those.

B4nan replied:
Same, unless we have an actual reason why we want to have it disabled?

jirimoravcik replied:
Ok. I don't see any actual reason for it being disabled.

B4nan commented on .github/workflows/publish.yaml:
```diff
@@ -0,0 +1,31 @@
+      -
+        uses: actions/setup-node@v4
+        with:
+          node-version: 20
```
nit: i'd go with 22, its already lts

B4nan commented on .github/workflows/publish.yaml:
```diff
@@ -0,0 +1,31 @@
+        uses: actions/setup-node@v4
+        with:
+          node-version: 20
+          registry-url: https://registry.npmjs.org/
```
this is most likely the default

B4nan commented on .github/workflows/publish.yaml:
```diff
@@ -0,0 +1,31 @@
+        run: npm install
+      -
+        name: Publish to NPM
+        run: NODE_AUTH_TOKEN=${{secrets.NPM_TOKEN}} npm publish --tag latest --access public
```
nit: maybe better to map the secret to env var separately instead of inlining it here. also `--tag latest` is the default.

B4nan commented on .github/workflows/publish.yaml:
```diff
@@ -0,0 +1,31 @@
+      -
+        name: Tag Version
+        run: |
+          git_tag=v`node -p "require('./package.json').version"`
+          git tag $git_tag
+          git push origin $git_tag
```
why not use `npm version` instead which will handle both git tag and bumping? this feels like you only deal with publishing and once released, things will fall apart unless someone bumps manually.

note that `npm version` should be used before `npm publish`

jirimoravcik replied:
the bumping should be done manually, yes.

B4nan replied:
I really dislike that, the new release process we are adopting these days don't suffer from this unnecessary manual step, why should we deal with this manually here? Basically all our public repos (except CLI which is in progress) don't requite this now, we release via workflow dispatch and bump based on conventional commits with possible manual override. We could adopt the same in here if you want, I'd let @janbuchar to make it work since its all his work. It also means automated changelogs and GH releases and other goodies.

jirimoravcik replied:
Ok, nice. Please send me some PR/commit that implements it, I'll do it here. Thanks!

B4nan replied:
One from the CLI, we have also canary builds in there, I don't know what we want from this repo, we could as well just ship to stable like we do in apify-shared-js, but if we want beta versions and stable releases via workflow dispatch, that's what the CLI and client are doing.

https://github.com/apify/apify-cli/pull/691

In client we had two PRs since we missed one thing

https://github.com/apify/apify-client-js/pull/596
https://github.com/apify/apify-client-js/pull/605

B4nan commented on README.md:
```diff
@@ -1,15 +1,30 @@
-# Usage
+# apify-eslint-config
```
(ocd)

```suggestion
# apify-eslint-config

```

B4nan commented on package.json:
```diff
@@ -1,24 +1,26 @@
+    "typescript-eslint": "*",
     "eslint": "*"
   },
+  "overrides": {
```
i thought `overrides` is for altering your local lockfile, which we don't have here at all, is this really working? i don't think it has effect on downstream

B4nan commented on index.js:
```diff
@@ -1,66 +1,187 @@
+            'import/extensions': 'off',
+            // Force ordering of imports.
+            'import/order': ['error', {
+                groups: ['builtin', 'external', ['parent', 'sibling'], 'index', 'object'],
+                alphabetize: {
+                    order: 'asc',
+                    caseInsensitive: true,
+                },
+                'newlines-between': 'always',
```
I am personally a bit allergic to this mixed quoting, but I guess it's better for not producing unnecessary git diffs... 

(this is not about the rules highlighted, but about the quoting on the keys)

jirimoravcik replied:
I don't follow

B4nan replied:
`alphabetize` is not quoted but `'newlines-between'` is

B4nan commented on index.js:
```diff
@@ -1,66 +1,187 @@
+            // Allow less strict formatting of function parentheses.
+            'function-paren-newline': 'off',
+            // Prefer destructuring for certain cases only. Forcing destructuring for arrays can make the code less readable.
+            'prefer-destructuring': ['error', {
```
(personal pet peeve)

If I could change one thing in how we write things in Apify, it would be this. We overuse (object) destructing so hard in here 😿

jirimoravcik replied:
Can you provide some examples, where you feel we overuse destructuring?

Hmm, wondering if the correct word is destructuring or destructing

B4nan replied:
All the open source repos had this enabled and use it extensively, same for actors. My problem with this is that you loose discoverability, as well as ability to mutate things. Wanna see what is available in actor context? You need to go to the function params instead of just writing `ctx.`. Wanna change some option we destruct on runtime? Sorry, cant do, since we introduced a new local var for it and used that instead. Also it feels like it results in more code while it was (imho) introduced to make us write less. Lastly, it forces people to do weird things like destructing `process.env`, which again suffers from the immutability, making tests of env vars much harder. 

But maybe its just me 🙃

Strajk replied:
> loose discoverability, as well as ability to mutate things

sorry for stalking but :this: 💯 👍

janbuchar replied:
I generally think that immutability is good, but the loss of discoverability is real. Reading the choose-your-adventure style code is a pain.

B4nan replied:
> Hmm, wondering if the correct word is destructuring or destructing

Oh lord, I have been calling it wrong the whole time, thanks for mentioning :D

B4nan commented on index.js:
```diff
@@ -1,66 +1,187 @@
+        },
+
+        rules: {
+            indent: ['error', 4, {
```
this is quite buggy (and deprecated) rule, it screws up with decorators (which is valid for the console nest app), they had quite a lot of issues about this, closed as wontfix, suggesting people to use formatters instead of it

jirimoravcik replied:
Yeah, formatting will be the next step, but it would make scope of the project too large, so I'll keep it here until we move forward with `prettier` in the future

B4nan replied:
yes, my point is we should not use this rule, not that we should add prettier right ahead. its broken and console team will have issues with that in the nest repo most likely.

B4nan commented on index.js:
```diff
@@ -1,66 +1,187 @@
+                SwitchCase: 1,
+            }],
+            // Allow underscore dangle for a few specific properties, mainly `_id`.
+            'no-underscore-dangle': [2, {
```
I would personally disable this one completely, we have it disabled in crawlee and will continue doing so, changing public props/methods (regardless of their `_` prefix`) is technically a BC. But I don't mind keeping it here and disabling on repository level if you think it's worth it, just be careful with this in libraries/shared packages.

jirimoravcik replied:
I'll keep it as it was in the config previously. I guess the repository level override makes sense here

B4nan reviewed (commented):
sorry for the late review, way too many adhoc things

B4nan commented on package.json:
```diff
@@ -1,24 +1,26 @@
+    "globals": "^15.11.0"
   },
   "peerDependencies": {
+    "typescript-eslint": "*",
```
maybe this should be optional peer dependency, e.g. for our JS actor templates it makes no sense. but I guess we should have a separate config for those anyway

jirimoravcik replied:
Makes sense to be optional, done.

jirimoravcik commented:
Hello @janbuchar I borrowed the nice release actions you made for Apify client. Can you check that they're setup alright here? I've done 2 changes: Removed wait for checks step and removed build step (before publish). Thanks

FYI the 2 secrets used are inherited from org so they should be available.

Also one question, will it behave ok since we don't have `package-lock.json` in this repo? and also no `CHANGELOG.md`?

janbuchar commented on .github/workflows/pre_release.yaml:
```diff
@@ -0,0 +1,94 @@
+            -   name: Update package version in package.json
+                run: npm version --no-git-tag-version --allow-same-version ${{ needs.release_metadata.outputs.version_number }}
+
+            -   name: Update CHANGELOG.md
+                uses: DamianReeves/write-file-action@master
+                with:
+                    path: CHANGELOG.md
+                    write-mode: overwrite
+                    contents: ${{ needs.release_metadata.outputs.changelog }}
```
If you don't want to keep a changelog for some reason, then you can delete this. Same applies to `release.yaml`.

janbuchar commented on .github/workflows/release.yaml:
```diff
@@ -0,0 +1,121 @@
+                uses: actions/setup-node@v4
+                with:
+                    node-version: 22
+                    cache: 'npm'
+                    cache-dependency-path: 'package-lock.json'
```
```suggestion
```

This shouldn't be necessary.

janbuchar commented on .github/workflows/pre_release.yaml:
```diff
@@ -0,0 +1,94 @@
+                uses: actions/setup-node@v4
+                with:
+                    node-version: 22
+                    cache: 'npm'
+                    cache-dependency-path: 'package-lock.json'
```
```suggestion
```
This shouldn't be necessary.

janbuchar reviewed (commented):
@jirimoravcik yeah, I believe it will work fine.
